custom/icds/rules/immunization.py

```diff
@@ -0,0 +1,216 @@
+"""
+This module contains utils which can be used for dealing with immunizations
+in the AWW app.
+
+Each immunization is represented by a CommCare Supply product. When an
+immunization is applied, it is recorded using a LedgerValue, where the entry_id
+of the LedgerValue is the product's unique id, and the balance of the
+LedgerValue is the date that the immunization took place, represented as an
+integer (the number of days since 1970-01-01). The LedgerValues are tracked
+against cases with case type "tasks".
+
+Each immunization has a window within which it is valid to be applied, which
+depends on a few factors, including whether it depends on another immunization
+to happen first. All of this information is stored as custom product_data on the
+CommCare Supply product.
+
+These utils can be used to calculate when immunizations are due for a given
+"tasks" case, whether that "tasks" case applies to a mother (case property
+"tasks_type" == "pregnancy"), or a child (case property "tasks_type" == "child").
+"""
+
+import pytz
+import re
+from corehq.apps.products.models import SQLProduct
+from corehq.form_processor.backends.sql.dbaccessors import LedgerAccessorSQL
+from corehq.form_processor.models import CommCareCaseIndexSQL
+from corehq.util.quickcache import quickcache
+from corehq.util.timezones.conversions import ServerTime
+from datetime import datetime, date, timedelta
+
+
+def _validate_tasks_case_and_immunization_product(tasks_case, immunization_product):
+    if tasks_case.get_case_property('tasks_type') != immunization_product.product_data.get('schedule'):
+        raise ValueError("Mismatch between tasks.tasks_type and product.schedule")
+
+
+def _validate_child_or_pregnancy_type(value):
+    if value not in ('child', 'pregnancy'):
+        raise ValueError("Expected one of: 'child', 'pregnancy'")
+
+
+@quickcache(['domain', 'schedule'], timeout=60 * 60)
+def get_immunization_products(domain, schedule):
+    _validate_child_or_pregnancy_type(schedule)
+
+    return [
+        product
+        for product in SQLProduct.active_objects.filter(domain=domain)
```

Comment:
How many products are expected to be in the domain? Could it be filtered by anything else like code or program id?

Reply:
there's less than 50 total in the domain at the moment. unfortunately the only other thing we could filter on is part of the custom product data which is not possible in the sql query (there are no programs defined). but i cached this result so this lookup in the db shouldn't happen often